lib/apis/v3/policy.go

```diff
@@ -59,13 +64,25 @@ type Rule struct {
 	Destination EntityRule `json:"destination,omitempty" validate:"omitempty"`
 
 	// HTTP contains match criteria that apply to HTTP requests.
-	HTTP *HTTPMatch `json:"http,omitempty" validate:"omitempty"`
+	HTTP *HTTPMatch `json:"http,omitempty" validate:"omitempty,dive"`
 }
 
+// HTTPMatch is an optional field that apply only to HTTP requests
+// The Methods and Path fields are joined with AND
 type HTTPMatch struct {
 	// Methods is an optional field that restricts the rule to apply only to HTTP requests that use one of the listed
 	// HTTP Methods (e.g. GET, PUT, etc.)
-	Methods []string `json:"methods,omitempty" validate:"omitempty"`
+	// Multiple methods are OR'd together.
+	Methods []string `json:"methods,omitempty" validate:"omitempty,httpMethods"`
+	// Paths is an optional field that restricts the rule to apply to HTTP requests that use one of the listed
+	// HTTP Paths. Paths may be either of the form:
+	// exact: <path> or
+	// prefix: <path-prefix>
+	// e.g:
+	// - exact: /foo
+	// - prefix: /bar
+	// Multiple paths are OR'd together.
+	Paths []map[string]string `json:"paths,omitempty" validate:"omitempty,httpPaths"`
```

Comment:
Why is this a generic `map[string]string` instead of defining a `HTTPPathMatch` struct?  You could structure it as 
```
type HTTPPathMatch struct {
    Path string
    MatchType string  // one of "prefix" "exact", defaults to exact
}
```

Reply:
Even if we don't change the format, it should be its own struct I think. e.g.

```
// Only one of Prefix or Exact can be specified (enforced by validation)
type HTTPMatch struct {
    Prefix string
    Exact string
}
```

Reply:
@caseydavenport since we desire the `oneof` (either prefix or exact) semantics. It would have to be an interface, I think.
Which is fine but then afaik the un-marshaling of a struct `HTTPMatch` which encloses an interface would have to be done manually by overriding the `json.UnmarshalJSON`.
 
For example:
```
type HTTPMatch struct {
   Methods []string `json:"methods,omitempty"`
   Paths []HTTPPath `json:"paths,omitempty"`
}

type HTTPPath interface {
  GetPath() string
}

type HTTPPathExact struct {
  Exact string `json:"exact"`
}

func (e *HTTPPathExact) GetPath() string { return e.Exact }

type HTTPPathPrefix struct {
  Prefix string `json:"prefix"`
}

func (e *HTTPPathPrefix) GetPath() string { return e.Prefix }

// HTTPMatch is enclosing struct for both HTTPMethod and HTTPPath
func (p *HTTPMatch) UnmarshalJSON(data []byte) error {
  var objMap map[string]*json.RawMessage

  err := json.Unmarshal(data, &objMap)
  if err != nil {
    fmt.Println("err1")
    return err
  }
....

  var rawPaths []*json.RawMessage
  err = json.Unmarshal(*objMap["paths"], &rawPaths)
  if err != nil {
    return err
  }

  f.Paths = make([]HTTPPath, len(rawPaths))

  var pathType map[string]string
  for index,  rawPath := range rawPaths {
    err := json.Unmarshal(*rawPath, &pathType)
    if err != nil {
      return err
    }
    fmt.Printf("Data: %#v\n", pathType)

    for k, v := range pathType {
      switch strings.ToLower(k) {
        case "exact":
          f.Paths[index] = &HTTPPathExact{Exact: v}
        case "prefix":
          f.Paths[index] = &HTTPPathPrefix{Prefix: v}
        default:
          return errors.New("Unsupported type for HTTPPath")

```

I feel that using an interface abstraction we are not gaining much and @fasaxc suggestion would be good middle ground between an implementation that is too loose (with a raw `map[string]string`) and having an API with a struct. Unless we feel that by having users specifying `MatchType` could be error prone.
wdyt?

Reply:
I don't understand why - just have both fields present in the same struct, and use validation to make sure the user never specified both of them. That is way less complicated than using interfaces. It's a common pattern in k8s APIs.

Reply:
I see what you mean! and will do so. 
@fasaxc Will make this changes and ping u when it's ready.